### ISSUE TITLE
Authentication Account Model review

### ISSUE THREAD

**Issue by djfarrelly** [bufferapp/authentication-service#13]
I'll just review this entire file right in this issue so it's all in one place 😄 

Should we declare these with `const`? they look like const's in ALL_CAPS, but if we don't declare I believe they become `var`'s and could be redefined.
https://github.com/bufferapp/authentication-service/blob/b8869b403ce6dcde7ecc3294eca387fc4f784111/src/authenticationAccountModel.js#L10-L11

---

What type of index is this? Is it a background index? I think it's background by default, but not sure! Can we confirm this or document this maybe?
**EDIT** - I confirmed it's a background index in https://github.com/bufferapp/authentication-service/issues/21#issuecomment-389332368

https://github.com/bufferapp/authentication-service/blob/b8869b403ce6dcde7ecc3294eca387fc4f784111/src/authenticationAccountModel.js#L15-L20

---

Should `hashedPassword` be required?
https://github.com/bufferapp/authentication-service/blob/b8869b403ce6dcde7ecc3294eca387fc4f784111/src/authenticationAccountModel.js#L21-L23

---

Removing `password` as a field. In my experience with Mongoose, adding a field as part of the schema just for validation purposes may not be the idiomatic way to do validation and things like this.

https://github.com/bufferapp/authentication-service/blob/b8869b403ce6dcde7ecc3294eca387fc4f784111/src/authenticationAccountModel.js#L24-L34

I did look into using Mongoose's VirtualType ([docs/guide](http://mongoosejs.com/docs/guide.html#virtuals)), but it seems since `bcrypt`'s `hash` method is async which, as you mentioned, is a positive thing so we can't use a virtual for this 😞. In this case, I'd recommend that we use a method like `setPassword` similar to the existing `verifyPassword` method. This also allows us to catch any error at the place that it is set rather than having it happen later in _"pre-save"_. 

```js
  authenticationAccountSchema.methods.setPassword = async function (password) {
    if (!password)
      throw new Error('Password is required')
    if (typeof password !== 'string')
      throw new Error('Password must be a String')
    if (password.length < MIN_PASSWORD_LENGTH)
      throw new Error(`Password must be at least ${MIN_PASSWORD_LENGTH} characters`)
    if (password.length > MAX_PASSWORD_LENGTH)
      throw new Error(`Password must be less than ${MAX_PASSWORD_LENGTH} characters`)
    this.hashedPassword = await hash(password, 10)
  }
```

```js
// updatePassword...
  await user.setPassword(newPassword)
  await user.save()
```

---

This is partially a style thing that I noticed. You're using object arguments in what seems like a way to use keyword arguments. I wonder with many of these methods only requiring a single argument if this adds extra verbosity to the code as the function name already highlights the name of the argument 🤔.

https://github.com/bufferapp/authentication-service/blob/b8869b403ce6dcde7ecc3294eca387fc4f784111/src/authenticationAccountModel.js#L75-L77
https://github.com/bufferapp/authentication-service/blob/b8869b403ce6dcde7ecc3294eca387fc4f784111/src/authenticationAccountModel.js#L80-L82
https://github.com/bufferapp/authentication-service/blob/b8869b403ce6dcde7ecc3294eca387fc4f784111/src/authenticationAccountModel.js#L87-L89

I do value this approach sometimes, and maybe you're thinking it keeps all method calls standard as objects? Part of me does feel that just a regular argument may be more simple and intuitive.

```js
user.verifyResetToken(resetToken)
```

I'm not too one way or another, but it is a style thing and I would like to hear if you think this is something to apply more broadly or share the benefits of this approach with other engineers. 

---

Another thing we chatted about already for the above methods was to drop `async` when the function itself was not asynchronous itself. I think this would mean dropping it on `verifyResetToken` and `verifyProductname`.

---

`verifyProductname` to `verifyProductName` to match the camelCase nature of the `productName` argument?

**Comment by hharnisc** [comment]
I've gone through and fixed these and moved everything over to a `setPassword` function.

On the last note about positional vs keyed arguments:

This is a personal style that comes from 2 observations related to readability and refactoring. Keyed arguments are much easier to come back and refactor, since position doesn't matter. Keyed arguments provide an extra hint about what an argument is for.

It might not be any time soon, but some day someone is going to inherit this code and need to read it and refactor it or extend it. What ends up happening eventually, with position arguments (even a single positional argument) is that someone extends a function with a boolean, int, or non descriptive string -- lets say for a concrete example we pick this function: 

https://github.com/bufferapp/authentication-service/blob/task/issue-13/src/authenticationAccountModel.js#L77

```js
  authenticationAccountSchema.methods.setPassword = async function({
    password,
  }) {
   // ... check stuff
    this.hashedPassword = await hash(password, 10)
  }
``` 
and the number of passes needs to be configured. So they make the number of passes an argument with a default:

```js
  authenticationAccountSchema.methods.setPassword = async function({
    password,
    passes = 10
  }) {
   // ... check stuff
    this.hashedPassword = await hash(password, passes)
  }
```
Now when someone uses the function with the passes variable it looks like this

```js
user.setPassword(password, 20)
```

When you read that code you'd have to look at the implementation of `setPassword` to understand what 20 does.

You might be thinking why not use a variable outside the function? That would be one solution, but it does not guarantee consistent documentation -- at least not in the same way as a keyed argument.

```js
user.setPassword({ password, passes: 20 })
```
TL;DR - this just increases the likely hood of people doing the right thing later, where the right thing is writing readable code.
